Migration guide for OData .NET 8 (#319)
MicrosoftDocs/OData-docs: Odata-docs/odatalib/whats-new/odatalib-8.md

- In OData .NET 8, nested inner errors are correctly serialized.
- OData .NET 8 renamed `ToJson()` method to `ToJsonString()`.
- When serializing an `ODataErrorDetail` instance, the OData serializers would write the optional target property ahead of the required message property. In OData .NET 8, we switched the order such that the required message property is written first.
- In OData .NET 7, `ODataError.ToString()` and `ODataJsonWriterUtils.WriteError(ODataError)` could produce different outputs in some cases. In OData .NET 8, they return the same output.
- In OData .NET 8, if a value cannot be serialized, an error is written in place of value for the corresponding property, instead of throwing an exception.

Here's a sample `ODataError` instance:

```csharp
var odataError = new ODataError
{
    Details = new Collection<ODataErrorDetail>
    {
        new ODataErrorDetail { Code = "OEDC1", Message = "OEDM1", Target = "OEDT1" },
        new ODataErrorDetail { Code = "OEDC1", Message = "OEDM1" },
        new ODataErrorDetail {}
    }
};
```

> Here's how it would be written in OData .NET 7 with `ODataJsonWriterUtils.WriteError(odataError, /* other parameters */)`:

Reviewer: Is this meant to be .NET 8

```json
{
  "error": {
    "code": "",
    "message": "",
    "target": "",
    "details": [
      {
        "errorcode": "OEDC1",
        "message": "OEDM1",
        "target": "OEDT1"
      },
      {
        "errorcode": "OEDC1",
        "message": "OEDM1",
        "target": ""
      },
      {
        "errorcode": "",
        "message": "",
        "target": ""
      }
    ],
    "innererror": {}
  }
}
```

Here's how it would be serialized in OData .NET 8 with either `ODataJsonWriterUtils.WriteError(odataError, /* other parameters */)` and `odataError.ToString()`:

```json
{
  "error": {
    "code": "",
    "message": "",
    "details": [
      {
        "code": "OEDC1",
        "message": "OEDM1",
        "target": "OEDT1"
      },
      {
        "code": "OEDC1",
        "message": "OEDM1"
      },
      {
        "code": "",
        "message": ""
      }
    ]
  }
}
```

### Changed `ODataLibraryCompatibility` into a flags enum

OData .NET 8 changed the `ODataLibraryCompatibility` enum into a flags enum, where each bit represents a different compatibility setting that can enable some legacy serialization behavior.

> In OData .NET 7, the enum had values like `ODataLibraryCompatibility.Version6`, `Version7` and `Latest`. These values were used to enable or disable certain behaviour based on desired compatibility with a certain library version after migration to a new version. In OData .NET 8, the we moved to a more granular option where you can set various compatibility options individually using bit flags, for example `ODataLibraryCompatibility.UseLegacyVariableCasing`, `WriteTopLevelODataNullAnnotation`, etc. This allows you to retain some legacy behaviour in the serializers when migrating to new versions. The enum still has `Version6` and `Version7` values, which now combine specific flags that are replicate the behaviour that was on by default for that specific version.

Reviewer: ```suggestion
In OData .NET 7, the enum had values like `ODataLibraryCompatibility.Version6`, `Version7` and `Latest`. These values were used to enable or disable certain behaviour based on desired compatibility with a certain library version after migration to a new version. In OData .NET 8, we moved to a more granular option where you can set various compatibility options individually using bit flags, for example `ODataLibraryCompatibility.UseLegacyVariableCasing`, `WriteTopLevelODataNullAnnotation`, etc. This allows you to retain some legacy behaviour in the serializers when migrating to new versions. The enum still has `Version6` and `Version7` values, which now combine specific flags that are replicate the behaviour that was on by default for that specific version.
```

### Serialize `variable` scale in lowercase

When writing the `Scale` attribute in XML CSDL, use `variable` in lowercase instead of `Variable`. An enum flag, `ODataLibraryCompatibility.UseLegacyVariableCasing`, was added to support the legacy behavior if desired.

### Changed JSON reader and writer interfaces

This is relevant if you implemented custom JSON reader or JSON writer interfaces.

> In OData .NET 7, there were many interfaces required to implement a custom JSON reader or write because the synchronous, asynchronous and streaming APIs were split.

Reviewer: ```suggestion
In OData .NET 7, there were many interfaces required to implement a custom JSON reader or writer because the synchronous, asynchronous and streaming APIs were split.
```

> In OData .NET 8 these different APIs have been consolidated. There's now only one `IJsonWriter` interface and only one `IJsonReader` interfaces that a customer writer or reader

Reviewer: ```suggestion
In OData .NET 8 these different APIs have been consolidated. There's now only one `IJsonWriter` interface and only one `IJsonReader` interface that a custom writer or reader
```

would need to implement.

OData .NET 7 also had two interfaces for JsonWriter factories, `IJsonWriterFactory` which accepted a `TextWriter` and `IStreamBasedJsonWriterFactory` accepted a `Stream`.
OData .NET 8 exposes a single factory interface, `IJsonWriterFactory` which accepts a `Stream`.

### Added `ODataMessageWriterSettings.ShouldIncludeAnnotation` for custom instance annotations

In OData .NET 7, custom instance annotations would not be included in the written output if they were not inluded in the `@odata.include-annotations` preference token in preference header.
However, in some cases, customers want to add custom instance annotations to the payload that should always be written regardless of end-user preference.

> In OData .NET 8, we introduced a `ShouldIncludeAnnotation` proeprty to the `ODataMessageWriterSettings` that allows developers to force a custom instance annotation to be written. It works as follows:

Reviewer 2: You made a typo in the word `Property`
```suggestion
In OData .NET 8, we introduced a `ShouldIncludeAnnotation` property to the `ODataMessageWriterSettings` that allows developers to force a custom instance annotation to be written. It works as follows:
```

- The property is a `Func<string, bool>` that receives an annotation name and returns true if the annotation should be written
- If the writer was going to skip a certain annotation (e.g. because it was not part of the annotations filter), it will check this delegate before skipping the annotation. If the delegate returns `true`, then the annotation will not be skipped.

It is important to note that returning false from the delegate does not guarantee that the annotation will not be written. For example, if an annotation was included in the preference header annotations filter, it may still be written even when the delegate returns false.

```csharp
writerSettings.ShouldIncludeAnnotation == (name) => name.StartsWith("important.");
```

### Changed `ODataBinaryStreamValue` to close underlying stream by default

In OData .NET 7, when `ODataBinaryStreamValue` class is initialized using the `ODataBinaryStreamValue(Stream)` constructor, the stream is left open by default upon the object being disposed.

> In OData .NET 8, the stream is closed by default when the object objects is disposed. `The ODataBinaryStreamValue(Stream, bool)` constructor overload may be used when leaving the stream open is intended.

Reviewer 2: I think you meant `the object is disposed` and not `the object objects is disposed` here.
```suggestion
In OData .NET 8, the stream is closed by default when the object is disposed. `The ODataBinaryStreamValue(Stream, bool)` constructor overload may be used when leaving the stream open is intended.
```

Author: Thanks.

In OData .NET 8, `Where` clause generates a `$filter` query even for key lookups by default. The `DataServiceContext.KeyComparisonGeneratesFilterQuery` setting is still available. It now defaults to `true`. You can restore the old behaviour by setting this flag to `false`. However, this setting has been marked as deprecated and could be removed in a future major version.

The recommended way to fetch an entity by ID is to use the `ByKey` method, e.g. `context.People.ByKey(10)`. This generates a URL to retrieve the entity by id (`/People(10)`). We encourage developers that use `Where` for key-lookups to switch to using `ByKey()` instead.

Using `$filter` when you intend to fetch an entity by key can lead to surprising results. For example, consider the following snippet of code:

```csharp
var context = new Container(new Uri("https://services.odata.org/TripPinRESTierService/"));
var friend = context.People.Where(p => p.UserName == "russellwhyte").Select(p => p.BestFriend);

Console.WriteLine(friend.FirstName);
```

> In OData .NET 7, the following OData URL would be generated by default: `https://services.odata.org/TripPinRESTierService/People('russellwhyte')/BestFriend`. Notice that the selected `BestFriend` is added as a navigation property segment to the URL. This request returns a the entity corresponding to the best friend at the top level of the response:

Reviewer: ```suggestion
In OData .NET 7, the following OData URL would be generated by default: `https://services.odata.org/TripPinRESTierService/People('russellwhyte')/BestFriend`. Notice that the selected `BestFriend` is added as a navigation property segment to the URL. This request returns the entity corresponding to the best friend at the top level of the response:
```

```json
{
    "@odata.context": "https://services.odata.org/TripPinRESTierService/(S(q5f0myjzjtfondm5l1a5qsbw))/$metadata#People/$entity",
    "UserName": "scottketchum",
    "FirstName": "Scott",
    "LastName": "Ketchum",
    ...
}
```

> However, in OData .NET 8, this code would generate the following URL: `https://services.odata.org/TripPinRESTierService/People?$filter=UserName eq 'russellwhyte'&$expand=BestFriend`. Notice that request filters the entity set collection and also includes the navigation property using the `$expand` query. This means the result will be a collection containing a single entity, which has an expanded `BestFriend` property:

Reviewer: ```suggestion
However, in OData .NET 8, this code would generate the following URL: `https://services.odata.org/TripPinRESTierService/People?$filter=UserName eq 'russellwhyte'&$expand=BestFriend`. Notice that the request filters the entity set collection and also includes the navigation property using the `$expand` query. This means the result will be a collection containing a single entity, which has an expanded `BestFriend` property:
```

Reviewer: Does `$filter` always return a collection even when only one item is found?

        {
            "UserName": "russellwhyte",
            "FirstName": "Russell",
            "LastName": "Whyte",
            ...
            "BestFriend": {
                "UserName": "scottketchum",
                "FirstName": "Scott",
                "LastName": "Ketchum",
                ...
            }
        }
    ]
}
```

This means that trying to access the result with `friend.FirstName` will fail and throw an exception, because the result is not the friend entity, but a collection that contains the filtered entity with the `BestFriend` property attached to it. Therefore in OData .NET 8 we would have to adjust the code as follows:

```csharp
// fetch people with expanded navigation property using OData URI People?$expand=BestFriend
var people = await context.People.Expand('BestFriend').ExecuteAsync();
// perform a local LINQ transform on the result data
var bestFriends = people.Select(p => p.BestFriend);

Console.WriteLine(bestFriends.First().FirstName);
```

Alternatively we could use the `ByKey` method:

```csharp
var friend = await ctx.People.ByKey("russelwhyte").Select(p => p.BestFriend).GetValueAsync();
Console.WriteLine(friend.FirstName);
```

> Alternatively, we could set `DataServiceContext.KeyComparisonGeneratesFilterQuery` to false, but it's no longer recommended:

Reviewer: I think we can get rid of this statement since we do not recommend this so that people get used to the other options.

Author: The reason I mentioned it here is for people for whom this change will break things, and migration might require changing a lot of code (if the old pattern was used in many places). This could provide a quick _temporary_ workaround.

```csharp
var context = new Container(new Uri("https://services.odata.org/TripPinRESTierService/"));
context.KeyComparisonGeneratesFilterQuery = false;

var friend = context.People.Where(p => p.UserName == "russellwhyte").Select(p => p.BestFriend);

Console.WriteLine(friend.FirstName);
```

### Dropped support for `HttpWebRequest`

OData .NET 8 removed support for `HttpWebRequest`, `HttpClient` should be used instead.

Subsequently the following related APIs have also been removed:

- The `HttpWebRequestMessage` class was removed.
- The `HttpRequestTransportMode` enum was removed.
- The `DataServiceContext.HttpRequestTransportMode` property was removed.

### Other obsolete APIs that have been removed

> - Obsolete `Credentials` property dropped from `DataServiceClientRequestMessage` abstract class. The recommended way to configure credentials is through `HttpClientHandler` that can be provided using `IHttpClientFactory`.

Reviewer: Is this necessary to provide a small snippet of how to configure this?

Author: I've added code samples in the section that talks about `IHttpClientFactory` support.
